cumulus/parachains/integration-tests/emulated/assets/asset-hub-rococo/src/tests/reserve_transfer.rs

```diff
@@ -302,138 +331,176 @@ fn limited_reserve_transfer_native_asset_from_system_para_to_para() {
+	let beneficiary_id = AssetHubRococoReceiver::get();
 	let amount_to_send: Balance = ASSET_HUB_ROCOCO_ED * 1000;
 	let assets = (Parent, amount_to_send).into();
 
 	let test_args = TestContext {
-		sender: AssetHubRococoSender::get(),
-		receiver: PenpalRococoAReceiver::get(),
-		args: system_para_test_args(destination, beneficiary_id, amount_to_send, assets, None),
+		sender: PenpalRococoASender::get(),
+		receiver: AssetHubRococoReceiver::get(),
+		args: para_test_args(destination, beneficiary_id, amount_to_send, assets, None, 0),
 	};
 
-	let mut test = SystemParaToParaTest::new(test_args);
+	let mut test = ParaToSystemParaTest::new(test_args);
 
 	let sender_balance_before = test.sender.balance;
+	let receiver_balance_before = test.receiver.balance;
+
+	let penpal_location_as_seen_by_ahr =
+		AssetHubRococo::sibling_location_of(PenpalRococoA::para_id());
+	let sov_penpal_on_ahr = AssetHubRococo::sovereign_account_id_of(penpal_location_as_seen_by_ahr);
 
-	test.set_assertion::<AssetHubRococo>(system_para_to_para_assertions);
-	// TODO: Add assertion for Penpal runtime. Right now message is failing with
-	// `UntrustedReserveLocation`
-	test.set_dispatchable::<AssetHubRococo>(system_para_to_para_reserve_transfer_assets);
+	// fund the Penpal's SA on AHR with the native tokens held in reserve
+	AssetHubRococo::fund_accounts(vec![(sov_penpal_on_ahr.into(), amount_to_send * 2)]);
+
+	test.set_assertion::<PenpalRococoA>(para_to_system_para_sender_assertions);
+	test.set_assertion::<AssetHubRococo>(para_to_system_para_receiver_assertions);
+	test.set_dispatchable::<PenpalRococoA>(para_to_system_para_limited_reserve_transfer_assets);
 	test.assert();
 
 	let sender_balance_after = test.sender.balance;
+	let receiver_balance_after = test.receiver.balance;
 
-	let delivery_fees = AssetHubRococo::execute_with(|| {
+	let delivery_fees = PenpalRococoA::execute_with(|| {
 		xcm_helpers::transfer_assets_delivery_fees::<
-			<AssetHubRococoXcmConfig as xcm_executor::Config>::XcmSender,
+			<PenpalRococoXcmConfig as xcm_executor::Config>::XcmSender,
 		>(test.args.assets.clone(), 0, test.args.weight_limit, test.args.beneficiary, test.args.dest)
 	});
 
+	// Sender's balance is reduced
 	assert_eq!(sender_balance_before - amount_to_send - delivery_fees, sender_balance_after);
-	// TODO: Check receiver balance when Penpal runtime is improved to propery handle reserve
-	// transfers
+	// Receiver's balance is increased
+	assert!(receiver_balance_after > receiver_balance_before);
 }
 
-/// Limited Reserve Transfers of a local asset from System Parachain to Parachain should work
+/// Reserve Transfers of a local asset and native asset from System Parachain to Parachain should
+/// work
 #[test]
-fn limited_reserve_transfer_asset_from_system_para_to_para() {
-	// Force create asset from Relay Chain and mint assets for System Parachain's sender account
+fn reserve_transfer_assets_from_system_para_to_para() {
```

Comment:
```
// Asset Hub may _act_ as a reserve location for ROC and assets created
// under `pallet-assets`
```
I am refering to the part `and assets created under pallet-assets`.

All emulated tests are using `impl_assets_helpers_for_parachain` which works with `pallet_assets::Instance1`, which is local trusted asset - `MultiLocation {parents: 0, interior: X1(GeneralIndex(xyz)}`.
I think we are missing test-case (at least emulated) for `ForeignAssets`.
We should add test for AssetHub runtimes which covers situation:
- in AssetHub's  `ForeignAssets` mint some ParaAToken - `MultiLocation {parents: 1, interior: X1(Parachain(ParaAId)}`  (mint is ok, because sibling parachains can teleport their tokens to AssetHub to `ForeignAssets`)
- in AssetHub's  `ForeignAssets` mint some ParaBToken - `MultiLocation {parents: 1, interior: X1(Parachain(ParaBId)}`
- `AssetHub::limited_reserve_transfer_assets(penpal_as_dest, [NativeTokenForFees, ParaAToken])`
- `AssetHub::limited_reserve_transfer_assets(penpal_as_dest, [NativeTokenForFees, ParaBToken])`
- `AssetHub::limited_reserve_transfer_assets(penpal_as_dest, [NativeTokenForFees, ParaAToken, ParaBToken])`
- check Penpal if assets are received

We need to add `pallet_assets` with `MultiLocation` support for `ForeignAssets` to the Penpal.

This test-case should work on actual master also, so we could/should also add this test as separate PR directly on master or here, wdyt?

Reply:
Yes, I already have this on another branch 🙈 and will add it separately after this PR. The reason I'm not including it here is because it is actually using teleports and not reserve transfers so it is orthogonal to this PR.

(AHs have teleport trust for foreign assets with their respective owner parachains: https://github.com/paritytech/polkadot-sdk/blob/ffa0e30e58234d85e24a3a55b088fb676ef71bfd/cumulus/parachains/runtimes/assets/asset-hub-rococo/src/xcm_config.rs#L562C40-L562C40)